lib/services/role.go

```diff
@@ -118,41 +132,330 @@ type Access interface {
+func (r *RoleV3) SetTTL(clock clockwork.Clock, ttl time.Duration) {
+	r.Metadata.SetTTL(clock, ttl)
+}
+
+// SetName sets the role name and is a shortcut for SetMetadata().Name.
+func (r *RoleV3) SetName(s string) {
+	r.Metadata.Name = s
+}
+
+// GetName gets the role name and is a shortcut for GetMetadata().Name.
+func (r *RoleV3) GetName() string {
+	return r.Metadata.Name
+}
+
+// GetMetadata returns role metadata.
+func (r *RoleV3) GetMetadata() Metadata {
+	return r.Metadata
+}
+
+// SetMaxSessionTTL sets the maximum duration for a SSH or Web session.
+func (r *RoleV3) SetMaxSessionTTL(duration time.Duration) {
+	r.Spec.MaxSessionTTL.Duration = duration
+}
+
+// GetMaxSessionTTL gets the maximum duration for a SSH or Web session.
+func (r *RoleV3) GetMaxSessionTTL() Duration {
+	return r.Spec.MaxSessionTTL
+}
+
+// GetOption gets an OpenSSH boolean option.
+func (r *RoleV3) GetOption(option string) bool {
+	return r.Spec.Options[option]
+}
+
+// SetOption sets an OpenSSH boolean option.
+func (r *RoleV3) SetOption(optionName string, optionValue bool) {
+	r.Spec.Options[optionName] = optionValue
+}
+
+// GetLogins gets system logins for allow or deny condition.
+func (r *RoleV3) GetLogins(rct RoleConditionType) []string {
+	if rct == Allow {
+		return r.Spec.Allow.Logins
+	}
+	return r.Spec.Deny.Logins
+}
+
+// SetLogins sets system logins for allow or deny condition.
+func (r *RoleV3) SetLogins(rct RoleConditionType, logins []string) {
+	if rct == Allow {
+		r.Spec.Allow.Logins = logins
+	} else {
+		r.Spec.Deny.Logins = logins
+	}
+}
+
+// GetNamespaces gets a list of namespaces this role is allowed or denied access to.
+func (r *RoleV3) GetNamespaces(rct RoleConditionType) []string {
+	if rct == Allow {
+		return r.Spec.Allow.Namespaces
+	}
+	return r.Spec.Deny.Namespaces
+}
+
+// GetNamespaces sets a list of namespaces this role is allowed or denied access to.
+func (r *RoleV3) SetNamespaces(rct RoleConditionType, namespaces []string) {
+	if rct == Allow {
+		r.Spec.Allow.Namespaces = namespaces
+	} else {
+		r.Spec.Deny.Namespaces = namespaces
+	}
+}
+
+// GetNodeLabels gets the map of node labels this role is allowed or denied access to.
+func (r *RoleV3) GetNodeLabels(rct RoleConditionType) map[string]string {
+	if rct == Allow {
+		return r.Spec.Allow.NodeLabels
+	}
+	return r.Spec.Deny.NodeLabels
+}
+
+// SetNodeLabels sets the map of node labels this role is allowed or denied access to.
+func (r *RoleV3) SetNodeLabels(rct RoleConditionType, labels map[string]string) {
+	if rct == Allow {
+		r.Spec.Allow.NodeLabels = labels
+	} else {
+		r.Spec.Deny.NodeLabels = labels
+	}
+}
+
+// GetRules gets all allow or deny rules.
+func (r *RoleV3) GetRules(rct RoleConditionType) []RoleRule {
+	if rct == Allow {
+		return r.Spec.Allow.Rules
+	}
+	return r.Spec.Deny.Rules
+}
+
+// SetRules sets an allow or deny rule.
+func (r *RoleV3) SetRule(rct RoleConditionType, resources []string, verbs []string) {
+	rr := RoleRule{
```

Comment:
I would make a full copy of a rule before storing it. In this case you can accept `resources` slice, which can actually be (probably by mistake) modified later by the caller, resulting in implicit modification of `RoleV2`  structure.